src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/AddConfigureAwaitAnalyzer.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed under the MIT License.
+
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Azure.ClientSdk.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class AddConfigureAwaitAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(Descriptors.AZC0011);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(AnalyzeAwaitExpression, SyntaxKind.AwaitExpression);
+        }
+
+        private void AnalyzeAwaitExpression(SyntaxNodeAnalysisContext context)
+        {
+            var invocation = (AwaitExpressionSyntax)context.Node;
+            var operation = context.SemanticModel.GetOperation(invocation, context.CancellationToken);
+            if (!(operation is IAwaitOperation awaitOperation)) {
+                return;
+            }
+
+            if (awaitOperation.Operation is IInvocationOperation configureAwaitOperation)
+            {
+                if (IsConfigureAwait(configureAwaitOperation.TargetMethod, context.Compilation))
+                {
+                    return;
+                }
+
+                if (!IsTaskType(configureAwaitOperation.Type, context.Compilation))
+                {
+                    return;
+                }
+            }
+
+            var location = awaitOperation.Syntax.GetLocation();
+            var diagnostic = Diagnostic.Create(Descriptors.AZC0011, location);
+            context.ReportDiagnostic(diagnostic);
+        }
+
+        private static bool IsConfigureAwait(IMethodSymbol method, Compilation compilation) 
+            => method.Name == "ConfigureAwait" && IsTaskType(method.ReceiverType, compilation);
+
+        private static bool IsTaskType(ITypeSymbol type, Compilation compilation) 
+        {
+            if (type == null) 
+            {
+                return false;
+            }
+
+            var (task, taskOfT, valueTask, valueTaskOfT) = GetTaskTypes(compilation);
+
+            if (task == null || taskOfT == null)
+            {
+                return false;
+            }
+
+            if (type.Equals(task) || type.Equals(valueTask)) 
+            {
+                return true;
+            }
+
+            var originalDefinition = type.OriginalDefinition;
+            if (Equals(originalDefinition, taskOfT) || Equals(originalDefinition, valueTaskOfT)) 
+            {
+                return true;
+            }
+
+            if (type.TypeKind == TypeKind.Error) 
+            {
+                return type.Name.Equals("Task") || type.Name.Equals("ValueTask");
+            }
```

Comment:
I assume this will happen when you can't resolve the type, but a comment would help clarify.